### PR TITLE
Betweenness centrality algorithms

### PR DESCRIPTION
**Pull request by mssemik**

### CODE REVIEW

**Comment by riomus** [comment]
Hi,

**Comment by mssemik** [comment]
Hi,
I will handle it at the weekend

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/hua/processor/NearlyOptimalBCProcessor.scala [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+      val sigma = expand.map(_.sigma).sum
+      val distance = expand.head.distance
+      val vertex = NOBFSVertex(distance, sigma, state = NOBFSVertex.toConfirm)
+      (root, vertex)
+    })
+
+    msgVertex2 ++ createdFlows
+  }
+
+  def sendMessages(round: Int)(ctx: EdgeContext[NOVertex, ED, List[NOMessage[VertexId]]]): Unit = {
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/hua/processor/NearlyOptimalBCProcessor.scala]: fixed

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/hua/processor/NearlyOptimalBCProcessor.scala [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+          val aggregations = messages.map(_.asInstanceOf[BCAggregationMessage].psi)
+          (root, flow.updateBC(aggregations))
+        case _ => throw new Error("Unsuspected case")
+      }
+    })
+
+    val createdFlows = expandMessages.map({ case (root, expand: List[BFSBCExtendMessage]) =>
+      if (msgVertex2.contains(root)) throw new Error("Attempt to create duplicate of vertex")
+      val sigma = expand.map(_.sigma).sum
+      val distance = expand.head.distance
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/hua/processor/NearlyOptimalBCProcessor.scala]: fixed

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    var i = 1
+
+    var g1 = cfbcProcessor.createFlow(initGraph).cache
+
+    var msg = cfbcProcessor.extractFlowMessages(g1).cache
+
+    var msgCount = msg.filter({ case (id, m) => m.nonEmpty }).count
+    var unfinalizedVertices = true
+
+    while (msgCount > 0 || unfinalizedVertices) {
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala]: loop condition depends on loop's body

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/edmonds/EdmondsBC.scala [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+  private def prepareRawGraph = {
+    graph.mapVertices((vId, attr) => EdmondsVertex())
+  }
+
+  def computeBC = {
+
+    val verticesIds = simpleGraph.vertices.map({ case (vertexId, _) => vertexId }).cache
+    val verticesIterator = verticesIds.toLocalIterator
+    var betweennessVector: VertexRDD[Double] = simpleGraph.vertices.mapValues(_ => .0).cache()
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/edmonds/EdmondsBC.scala]: BC values are updated after each algorithm step

File: operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                                                 send: (Int) => (EdgeContext[VD, ED, MD]) => Unit,
+                                                 merge: (MD, MD) => MD,
+                                                 interruptAfterEmptyRounds: Int = 1) = {
+
+
+    var round = 0
+    var numOfEmptyRounds = 0
+
+    var g = graph.mapVertices((vid, vdata) => prepareVertices(vdata))
+    var messages = g.aggregateMessages(send(round), merge)
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala]: values updated in each loop step

File: operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    var activeMessages = messages.count()
+
+    var prevG: Option[Graph[VD, ED]] = None
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala]: variable used to keep values to unpersist in next iteration

File: operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  def apply[OVD, VD: ClassTag, ED, MD: ClassTag](graph: Graph[OVD, ED],
+                                                 prepareVertices: (OVD) => VD,
+                                                 vpred: (Int) => (VertexId, VD, Option[MD]) => VD,
+                                                 send: (Int) => (EdgeContext[VD, ED, MD]) => Unit,
+                                                 merge: (MD, MD) => MD,
+                                                 interruptAfterEmptyRounds: Int = 1) = {
+
+
+    var round = 0
+    var numOfEmptyRounds = 0
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala]: counter of rounds with no messages (need in one of BC algorithm where rounds with no messages is likely)

File: operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    var g = graph.mapVertices((vid, vdata) => prepareVertices(vdata))
+    var messages = g.aggregateMessages(send(round), merge)
+    var activeMessages = messages.count()
+
+    var prevG: Option[Graph[VD, ED]] = None
+    while (activeMessages > 0 || numOfEmptyRounds < interruptAfterEmptyRounds) {
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala]: any other loop doesn't match in this case

Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala]: other loop doesn't match in this case

File: operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+  def apply[OVD, VD: ClassTag, ED, MD: ClassTag](graph: Graph[OVD, ED],
+                                                 prepareVertices: (OVD) => VD,
+                                                 vpred: (Int) => (VertexId, VD, Option[MD]) => VD,
+                                                 send: (Int) => (EdgeContext[VD, ED, MD]) => Unit,
+                                                 merge: (MD, MD) => MD,
+                                                 interruptAfterEmptyRounds: Int = 1) = {
+
+
+    var round = 0
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala]: counter of round used for checkpoint values of RDD

File: operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    var numOfEmptyRounds = 0
+
+    var g = graph.mapVertices((vid, vdata) => prepareVertices(vdata))
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala]: graph is modified in ech loop step

File: operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    var messages = g.aggregateMessages(send(round), merge)
+    var activeMessages = messages.count()
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/algorithms/pregel/Pregel.scala]: loop condition updated in loops body

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala (hunk 3 above)
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala]: other loop doesn't match in this case

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+      val sample = Random.shuffle(m.distinct.diff(List(id)))
+      CFBCVertex(id, v.degree, v.bc, sample.toArray)
+
+    })
+
+    var i = 1
+
+    var g1 = cfbcProcessor.createFlow(initGraph).cache
+
+    var msg = cfbcProcessor.extractFlowMessages(g1).cache
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala]: messages change in each iteration

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    val ctrw = new ContinuousTimeRandomWalk[CFBCVertex, ED](cfbcProcessor.initGraph)
+    val randomVertices = ctrw.sampleVertices(Math.ceil(Math.max(1, k / 10)) toInt)
+
+    val initGraph = cfbcProcessor.initGraph.ops.joinVertices(randomVertices)((id, v, m) => {
+      val sample = Random.shuffle(m.distinct.diff(List(id)))
+      CFBCVertex(id, v.degree, v.bc, sample.toArray)
+
+    })
+
+    var i = 1
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala]: loop counter used for checkpointing rdd

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+    var g1 = cfbcProcessor.createFlow(initGraph).cache
+
+    var msg = cfbcProcessor.extractFlowMessages(g1).cache
+
+    var msgCount = msg.filter({ case (id, m) => m.nonEmpty }).count
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala]: # of messages change in each iteration

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+    val initGraph = cfbcProcessor.initGraph.ops.joinVertices(randomVertices)((id, v, m) => {
+      val sample = Random.shuffle(m.distinct.diff(List(id)))
+      CFBCVertex(id, v.degree, v.bc, sample.toArray)
+
+    })
+
+    var i = 1
+
+    var g1 = cfbcProcessor.createFlow(initGraph).cache
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala]: temporary graph updated in each iteration

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+    var msg = cfbcProcessor.extractFlowMessages(g1).cache
+
+    var msgCount = msg.filter({ case (id, m) => m.nonEmpty }).count
+    var unfinalizedVertices = true
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/flow/CurrentFlowBC.scala]: part of loop condition

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/edmonds/EdmondsBCAggregator.scala [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+  * Created by mth on 4/12/17.
+  */
+class EdmondsBCAggregator[ED] extends Serializable {
+
+  def aggregate(graph: Graph[EdmondsVertex, ED], source: VertexId) = {
+//    val startTime = System.nanoTime()
+
+    val maxDepth = graph.vertices.aggregate(0)({ case (depth, (vId, vData)) => Math.max(vData.depth, depth) }, Math.max)
+
+    var g = graph
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/edmonds/EdmondsBCAggregator.scala]: temporary graph updated in each iteration

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/edmonds/EdmondsBCAggregator.scala [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    var oldGraph: Option[Graph[EdmondsVertex, ED]] = None
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/edmonds/EdmondsBCAggregator.scala]: graph from prev iteration (keep for unpersist in next iteration)

File: operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/edmonds/EdmondsBCAggregator.scala [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+    var messages = aggregateMessages(g, maxDepth).cache
```
Comment (mssemik) [inline operators/src/main/scala/ml/sparkling/graph/operators/measures/vertex/betweenness/edmonds/EdmondsBCAggregator.scala]: messages change in each iteration

**Comment by riomus** [comment]
Hi, I see that you are commenting a lot of issues and also fixing them. Nice, are you going to continue that trend? Those functionalities look really promising and are important for the development of the whole library.

**Comment by mssemik** [comment]
Hi, I was looking for better solution for mutable local variables but I didn't find any alternative solution in those cases yet. If you have any hits for me how can I make my code better, I will be glad. If not, feel free to complete pull request.

**Pull request merged**